execution/executor/src/chunk_executor.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    components::{
+        chunk_commit_queue::ChunkCommitQueue, chunk_output::ChunkOutput,
+        executed_chunk::ExecutedChunk,
+    },
+    logging::{LogEntry, LogSchema},
+    metrics::{
+        DIEM_EXECUTOR_APPLY_CHUNK_SECONDS, DIEM_EXECUTOR_COMMIT_CHUNK_SECONDS,
+        DIEM_EXECUTOR_EXECUTE_CHUNK_SECONDS,
+    },
+};
+use anyhow::Result;
+use diem_infallible::Mutex;
+use diem_logger::prelude::*;
+use diem_state_view::StateViewId;
+use diem_types::{
+    contract_event::ContractEvent,
+    ledger_info::LedgerInfoWithSignatures,
+    protocol_spec::DpnProto,
+    transaction::{
+        default_protocol::{TransactionListWithProof, TransactionOutputListWithProof},
+        TransactionInfo,
+    },
+};
+use diem_vm::VMExecutor;
+use executor_types::{ChunkExecutorTrait, ExecutedTrees};
+use fail::fail_point;
+use std::{marker::PhantomData, sync::Arc};
+use storage_interface::{default_protocol::DbReaderWriter, state_view::VerifiedStateView};
+
+pub struct ChunkExecutor<V> {
+    db: DbReaderWriter,
+    commit_queue: Mutex<ChunkCommitQueue>,
```

Comment:
Yes, 

>  every time it detects that the queue is not empty
and the current design is every time a chunk is finished execution and enqueued, the execution thread/task need to inform the commission thread/task to call commit_chunk() ONCE (it errors out if the queue is empty)

Reply:
I'm curious why you think this model is cleaner? In my mind, it's almost like a publish subscribe queue, right? We have a publisher (execute) and a subscriber (commit)? From the client perspective, it seems odd that we'd require the publisher to inform the subscriber out-of-band. Anyway, it's a small difference -- don't feel super strongly 😄

Reply:
The state sync and the db-restore can choose to drive the pipeline differently. db-restore use async streams and Tokio tasks, doesn't need to explicitly wait for the last chunk commission, etc; and the state sync might have a commission thread spawned, and it might need to know explicitly if the commission queue is drained, etc. 

The ChunkExecutor implements the data structure (an ExecutedChunk queue), and identify the critical section for making sure it's thread safe (the mutex and mutex inside the SMT). On top of it we can have different ways to drive the parallelism at runtime. If both drive it the same way, we can extract that to a common part as well.

Ok to be honest I just like smaller structures (for me to build). I agree we can be pretty flexible and change at will any time.

Reply:
Makes sense, thanks @msmouse 😄